File: src/bin/cli/cmd/get.rs
use crate::cli::OutputFormat;
use crate::dto::SubtitleDto;
use crate::json_output::{output_error, output_success};
use crate::utils;
use lib::subtitle::ports::SubtitleService;
use lib::subtitle::service::SubRipService;
use log::{debug, info};
use serde::Serialize;

#[derive(Serialize)]
struct GetRangeDto {
    subtitles: Vec<SubtitleDto>,
    count: usize,
    range_start: u32,
    range_end: u32,
}

pub fn handle(file: &str, index_str: &str, format: &OutputFormat) -> anyhow::Result<()> {
    if index_str.contains('-') {
        debug!("detected range format, processing range");
        return handle_range(file, index_str, format);
    }

    let index = index_str.parse::<u32>().map_err(|_| {
        anyhow::anyhow!(
            "Invalid index '{}'. Must be a positive number or range (e.g., 120-123)",
            index_str
        )
    })?;

    info!("getting subtitle {} from file: {}", index, file);

    let resolved = utils::resolve_existing_path(file)?;
    let service = SubRipService::new(resolved.base_dir);

    debug!("retrieving subtitle by id: {}..", index);
    match service.get_by_id(&resolved.filename, index) {
        Ok(Some(subtitle)) => {
            info!("subtitle {} found successfully", index);
            let dto = SubtitleDto::from_subtitle(&subtitle);
            output_success(format, &dto, || {
                println!("{}", subtitle);
            });
            Ok(())
        }
        Ok(None) => {
            output_error(
                format,
                "subtitle_not_found",
                &format!("subtitle with index {} not found in {}", index, file),
                None,
            );
            Err(anyhow::anyhow!(""))
        }
        Err(e) => {
            let cli_err = utils::format_subtitle_error(&e, file);
            output_error(
                format,
                &cli_err.code,
                &cli_err.message,
                cli_err.hint.as_deref(),
            );
            Err(anyhow::anyhow!(""))
        }
    }
}

fn handle_range(file: &str, range: &str, format: &OutputFormat) -> anyhow::Result<()> {
    info!("getting subtitles in range {} from file: {}", range, file);

    let (start, end) = utils::parse_range(range)?;
    debug!("parsed range: start={}, end={}", start, end);

    let resolved = utils::resolve_existing_path(file)?;
    let service = SubRipService::new(resolved.base_dir);

    debug!("retrieving all subtitles for filtering..");
    match service.get_all(&resolved.filename) {
        Ok(subtitles) => {
            debug!("found {} total subtitles", subtitles.len());

            let range_subtitles: Vec<_> = subtitles
                .into_iter()
                .filter(|s| {
                    let index = *s.index.as_ref();
                    index >= start && index <= end
                })
                .collect();

            debug!(
                "found {} subtitle(s) in range {}-{}",
                range_subtitles.len(),
                start,
                end
            );

            if range_subtitles.is_empty() {
                info!("no subtitles found in range {}-{}", start, end);
                let dto = GetRangeDto {
                    subtitles: vec![],
                    count: 0,
                    range_start: start,
                    range_end: end,
                };
                output_success(format, &dto, || {
                    println!("No subtitles found in range {}-{}", start, end);
                });
            } else {
                info!(
                    "found {} subtitle(s) in range {}-{}",
                    range_subtitles.len(),
                    start,
                    end
                );

                let dto = GetRangeDto {
                    subtitles: range_subtitles
                        .iter()
                        .map(SubtitleDto::from_subtitle)
                        .collect(),
                    count: range_subtitles.len(),
                    range_start: start,
                    range_end: end,
                };

                output_success(format, &dto, || {
                    for (i, subtitle) in range_subtitles.iter().enumerate() {
                        print!("{}", subtitle);
                        if i < range_subtitles.len() - 1 {
                            println!("\n");
                        } else {
                            println!();
                        }
                    }
                });
            }

            Ok(())
        }
        Err(e) => {
            let cli_err = utils::format_subtitle_error(&e, file);
            output_error(
                format,
                &cli_err.code,
                &cli_err.message,
                cli_err.hint.as_deref(),
            );
            Err(anyhow::anyhow!(""))
        }
    }
}

File: internal/cli/cmd/get.go
package cmd

import (
	"errors"
	"fmt"
	"log/slog"
	"strconv"
	"strings"

	"github.com/lebe-dev/submarine/internal/cli"
	"github.com/lebe-dev/submarine/pkg/subtitle"
)

// getRangeDto is a port of the Rust `struct GetRangeDto`.
type getRangeDto struct {
	Subtitles  []cli.SubtitleDto `json:"subtitles"`
	Count      int               `json:"count"`
	RangeStart uint32            `json:"range_start"`
	RangeEnd   uint32            `json:"range_end"`
}

// HandleGet is the port of the Rust `cmd::get::handle`.
func HandleGet(file, indexStr string, format cli.OutputFormat) error {
	if strings.Contains(indexStr, "-") {
		slog.Debug("detected range format, processing range")
		return handleGetRange(file, indexStr, format)
	}

	index, err := strconv.ParseUint(indexStr, 10, 32)
	if err != nil {
		return fmt.Errorf(
			"Invalid index '%s'. Must be a positive number or range (e.g., 120-123)",
			indexStr,
		)
	}
	idx := uint32(index)

	slog.Info(fmt.Sprintf("getting subtitle %d from file: %s", idx, file))

	resolved, err := cli.ResolveExistingPath(file)
	if err != nil {
		return err
	}
	service := subtitle.NewSubRipService(resolved.BaseDir)

	slog.Debug(fmt.Sprintf("retrieving subtitle by id: %d..", idx))
	sub, err := service.GetByID(resolved.Filename, idx)
	if err != nil {
		cliErr := getFormatSubtitleError(err, file)
		cli.OutputError(format, cliErr.Code, cliErr.Message, cliErr.Hint)
		return errors.New("")
	}

	if sub == nil {
		cli.OutputError(
			format,
			"subtitle_not_found",
			fmt.Sprintf("subtitle with index %d not found in %s", idx, file),
			nil,
		)
		return errors.New("")
	}

	slog.Info(fmt.Sprintf("subtitle %d found successfully", idx))
	dto := cli.SubtitleDtoFromSubtitle(sub)
	cli.OutputSuccess(format, dto, func() {
		fmt.Println(sub.String())
	})
	return nil
}

// handleGetRange is the port of the Rust `cmd::get::handle_range`.
func handleGetRange(file, rng string, format cli.OutputFormat) error {
	slog.Info(fmt.Sprintf("getting subtitles in range %s from file: %s", rng, file))

	start, end, err := cli.ParseRange(rng)
	if err != nil {
		return err
	}
	slog.Debug(fmt.Sprintf("parsed range: start=%d, end=%d", start, end))

	resolved, err := cli.ResolveExistingPath(file)
	if err != nil {
		return err
	}
	service := subtitle.NewSubRipService(resolved.BaseDir)

	slog.Debug("retrieving all subtitles for filtering..")
	subtitles, err := service.GetAll(resolved.Filename)
	if err != nil {
		cliErr := getFormatSubtitleError(err, file)
		cli.OutputError(format, cliErr.Code, cliErr.Message, cliErr.Hint)
		return errors.New("")
	}

	slog.Debug(fmt.Sprintf("found %d total subtitles", len(subtitles)))

	rangeSubtitles := make([]subtitle.Subtitle, 0)
	for i := range subtitles {
		idx := subtitles[i].Index.Value()
		if idx >= start && idx <= end {
			rangeSubtitles = append(rangeSubtitles, subtitles[i])
		}
	}

	slog.Debug(fmt.Sprintf(
		"found %d subtitle(s) in range %d-%d",
		len(rangeSubtitles), start, end,
	))

	if len(rangeSubtitles) == 0 {
		slog.Info(fmt.Sprintf("no subtitles found in range %d-%d", start, end))
		dto := getRangeDto{
			Subtitles:  []cli.SubtitleDto{},
			Count:      0,
			RangeStart: start,
			RangeEnd:   end,
		}
		cli.OutputSuccess(format, dto, func() {
			fmt.Printf("No subtitles found in range %d-%d\n", start, end)
		})
		return nil
	}

	slog.Info(fmt.Sprintf(
		"found %d subtitle(s) in range %d-%d",
		len(rangeSubtitles), start, end,
	))

	dtos := make([]cli.SubtitleDto, 0, len(rangeSubtitles))
	for i := range rangeSubtitles {
		dtos = append(dtos, cli.SubtitleDtoFromSubtitle(&rangeSubtitles[i]))
	}

	dto := getRangeDto{
		Subtitles:  dtos,
		Count:      len(rangeSubtitles),
		RangeStart: start,
		RangeEnd:   end,
	}

	cli.OutputSuccess(format, dto, func() {
		for i := range rangeSubtitles {
			fmt.Print(rangeSubtitles[i].String())
			if i < len(rangeSubtitles)-1 {
				fmt.Print("\n\n")
			} else {
				fmt.Println()
			}
		}
	})
	return nil
}

// getFormatSubtitleError unwraps a service error into a *subtitle.SubtitleError
// and runs it through cli.FormatSubtitleError, mirroring the Rust
// `utils::format_subtitle_error` call.
func getFormatSubtitleError(err error, file string) cli.CliError {
	var subErr *subtitle.SubtitleError
	if errors.As(err, &subErr) {
		return cli.FormatSubtitleError(subErr, file)
	}
	return cli.CliError{Code: "io_error", Message: err.Error()}
}
